src/main/java/org/alfresco/repo/domain/schema/script/DeleteNotExistsExecutor.java

```diff
@@ -0,0 +1,555 @@
+//        }
+
+        Statement stmt = connection.createStatement();
+        Object ret = null;
+        try {
+            if (logger.isDebugEnabled()) {
+                LogUtil.debug(logger, MSG_EXECUTING_STATEMENT, sql);
+            }
+            boolean haveResults = stmt.execute(sql);
+            // Record the statement
+//            executedStatements.append(sql).append(";\n\n");
+            if (haveResults && fetchColumnName != null) {
+                ResultSet rs = stmt.getResultSet();
+                if (rs.next()) {
+                    // Get the result value
+                    ret = rs.getObject(fetchColumnName);
+                }
+            }
+        } catch (SQLException e) {
+            if (optional) {
+                // it was marked as optional, so we just ignore it
+                LogUtil.debug(logger, MSG_OPTIONAL_STATEMENT_FAILED, sql, e.getMessage(), file.getAbsolutePath(), line);
+            } else {
+                LogUtil.error(logger, ERR_STATEMENT_FAILED, sql, e.getMessage(), file.getAbsolutePath(), line);
+                throw e;
+            }
+        } finally {
+            try {
+                stmt.close();
+            } catch (Throwable e) {
+            }
+        }
+        return ret;
+    }
+
+    private int getBatchUpperLimit(Connection connection, String tableName, String columnName, int line, File scriptFile) throws Exception {
+        int batchUpperLimit = 0;
+
+        String stmt = "SELECT MAX(" + columnName + ") AS upper_limit FROM " + tableName;
+        Object fetchedVal = executeStatement(connection, stmt, "upper_limit", false, line, scriptFile);
+
+        if (fetchedVal instanceof Number) {
+            batchUpperLimit = ((Number) fetchedVal).intValue();
+        }
+
+        return batchUpperLimit;
+    }
+
+    private boolean isLess(Long primaryId, Long[] secondaryIds) {
+        for (Long secondaryId : secondaryIds) {
+            if (secondaryId != null && primaryId >= secondaryId) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private String createPreparedSelectStatement(String tableName, String columnName)
+    {
+        return "SELECT " + columnName + " FROM " + tableName + " WHERE " + columnName + " > ? AND " + columnName + " <= ? ORDER BY " + columnName + " ASC";
+    }
+
+    private void createAndExecuteDeleteStatement(Connection connection, String tableName, String idColumnName, Set<Long> deleteIds, int maxBatchSize, int line, File scriptFile) throws Exception {
+        if (deleteIds.isEmpty()) {
+            return;
+        }
+
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("Prepare to delete " + deleteIds.size() + " items from table " + tableName + ".");
+        }
+
+        StringBuilder stmtBuilder = new StringBuilder("DELETE FROM " + tableName + " WHERE " + idColumnName + " IN ");
```

Comment:
It is possible to do this with a prepared statement too. Generally half the processing time is preparing, so is worth doing if it is run many times. Very long statements may also be an issue with some DBs. Where you don't have a full set of IDs for your prepared statement, you would just repeat one of the IDs many times when you bind variables to the prepared statement.

Reply:
Changes done